web/pandas/pdeps/0009-io-extensions.md

```diff
@@ -0,0 +1,395 @@
+and representations they are interested in, in an easy and intuitive way.
+
+Third-party packages are already able to implement connectors to pandas, but
+there are some limitations to it:
+
+- Given the large number of formats supported by pandas itself, third-party
+  connectors are likely seen as second class citizens, not important enough
+  to be used, or not well supported.
+- There is no standard API for external I/O connectors, and users need
+  to learn each of them individually.
+- Method chaining is not possible with third-party I/O connectors to export
+  data, unless authors monkey patch the `DataFrame` class, which should not
+  be encouraged.
+
+This document proposes to open the development of pandas I/O connectors to
+third-party libraries in a standard way that overcomes those limitations.
+
+### Proposal implementation
+
+Implementing this proposal would not require major changes to pandas, and
+the API defined next would be used.
+
+#### User API
+
+Users will be able to install third-party packages implementing pandas
+connectors using the standard packaging tools (pip, conda, etc.). These
+connectors should implement an entrypoint that pandas will use to
+automatically create the corresponding methods `pandas.read_*` and
+`pandas.DataFrame.to_*`. Arbitrary function or method names will not
+be created by this interface, only the `read_*` and `to_*` pattern will
+be allowed. By simply installing the appropriate packages users will
+be able to use code like this:
+
+```python
+import pandas
+
+df = pandas.read_duckdb("SELECT * FROM 'dataset.parquet';")
+
+df.to_hive(hive_conn, "hive_table")
+```
+
+This API allows for method chaining:
+
+```python
+(pandas.read_duckdb("SELECT * FROM 'dataset.parquet';")
+       .to_hive(hive_conn, "hive_table"))
+```
+
+The total number of I/O functions and methods is expected to be small, as users
+in general use only a small subset of formats. The number could actually be
+reduced from the current state if the less popular formats (such as SAS, SPSS,
+BigQuery, etc.) are removed from the pandas core into third-party packages.
+Moving these connectors is not part of this proposal, and will be discussed
+later in a separate proposal.
+
+#### Plugin registration
+
+Third-party packages would implement an
+[entrypoint](https://setuptools.pypa.io/en/latest/userguide/entry_point.html#entry-points-for-plugins)
+to define the connectors that they implement, under a group `dataframe.io`.
+
+For example, a hypothetical project `pandas_duckdb` implementing a `read_duckdb`
+function, could use `pyproject.toml` to define the next entry point:
+
+```toml
+[project.entry-points."dataframe.io"]
+from_duckdb = "pandas_duckdb:read_duckdb"
+```
+
+On import of the pandas module, it would read the entrypoint registry for the
+`dataframe.io` group, and would dynamically create methods in the `pandas` and
+`pandas.DataFrame` namespace for them. Method not starting with `read_` or `to_`
```

Comment:
`pandas.Series` should also have `to_` methods too?

Reply:
I'm open to include it here, but I'd personally leave `Series` out of this proposal. If later we see it's useful, we can surely review this proposal or create a new one for it. But personally, I don't see too many use cases, the current `Series.to_` methods are mostly internal (`to_period`, `to_timestap`, besides `to_numpy` and `to_list`), and it can be confusing adding third-party things to those in my opinion.

As said, I'm surely open to add it if there is interest and there are actual uses cases, but I wouldn't include it in this proposal.

Reply:
Ok, sorry, I just realized now that we have `to_csv`, `to_xarray` and most of them also for `Series`, never used one of those. I got confused because in the API documentation the internal ones and the rest are not together...

Do you think it can make sense that the connectors don't know anything about `Series`, and our `Series` methods simply call `to_frame()` before calling the connector? Or do you see important cases where there should be an important distinction between a Series and a one-column DataFrame? I can think of JSON as a case where there would be a subtle difference, but not much more.